### ISSUE TITLE
Add support for COPY

### ISSUE THREAD

**Issue by seriyps** [epgsql/epgsql#137]
Protocol description
https://www.postgresql.org/docs/current/static/protocol-flow.html#PROTOCOL-COPY
Protocol packet description
https://www.postgresql.org/docs/current/static/protocol-message-formats.html

We should also think on API:

Should we accept `file:io_device()`? Just raw binary() / iolist() data? Filenames `file:name_all()`.

**Comment by davidw** [comment]
Is there a use case driving this, or is it just a general "we should do this"?

**Comment by seriyps** [comment]
I personally don't have use cases for it, but I just noticed that most of the PG drivers supports this. Eg,
elixir: https://hexdocs.pm/postgrex/Postgrex.html#stream/4
Python: http://initd.org/psycopg/docs/usage.html#using-copy-to-and-copy-from
Rust: https://docs.rs/postgres/0.15.1/postgres/stmt/struct.Statement.html#method.copy_in
NodeJS: https://github.com/brianc/node-pg-copy-streams
Go (partialy): https://github.com/lib/pq#future--things-you-can-help-with

So yep, it's more close to "why not?"

**Comment by GuillaumeGandon** [comment]
> Is there a use case driving this, or is it just a general "we should do this"?

From [PostgreSQL documentation](https://www.postgresql.org/docs/current/populate.html) it is the fastest method to populate a DB:

1. > Note that loading a large number of rows using COPY is **almost always faster** than using INSERT, even if PREPARE is used and multiple insertions are batched into a single transaction.

1. > If you are loading a freshly created table, the **fastest method** is to create the table, bulk load the table's data using COPY, then create any indexes needed for the table.

**Comment by dszoboszlay** [comment]
> Is there a use case driving this, or is it just a general "we should do this"?

I have a very specific use case for this feature, as I'd like to convert a lot of data from a different DB to Postgres.

> We should also think on API:

The DB conversion goes like: read a chunk of data from one DB, maybe make some schema adjustments in memory and then dump to Postgres. This would need some streaming API from `epgsql`, because it's not feasible to hold the entire DB to be converted in memory. Dumping it to disk and reading back is a waste of time too.

It would be also useful if `epgsql` could do the conversion between Erlang types and PG types as usual.

**Comment by seriyps** [comment]
I was thinking about implementing it as an Erlang's [io protocol](https://erlang.org/doc/apps/stdlib/io_protocol.html).

So, smth like
```erlang
{ok, IoDescriptor} = epgsql:start_copy(C, "COPY my_table FROM stdin ..."),
ok = file:write(IoDescriptor, <<"1, 'hello world'\n">>),
io:format(IoDescriptor, "blabla~pblabla\n", [Args]),
ok = file:close(IoDescriptor),
epgsql:squery(C, "SELECT * FROM my_table").
```
etc.

So, you could write arbitrary CSV-formatted data using `file` module's functions.

Please, let me know if you have other proposals? I don't have much experience with using `COPY`, so, not sure what the most comfortable API would look like.

**Comment by GuillaumeGandon** [comment]
> I was thinking about implementing it as an Erlang's [io protocol](https://erlang.org/doc/apps/stdlib/io_protocol.html).
> 
> So, smth like
> 
> ```erlang
> {ok, IoDescriptor} = epgsql:start_copy(C, "COPY my_table FROM stdin ..."),
> ok = file:write(IoDescriptor, <<"1, 'hello world'\n">>),
> io:format(IoDescriptor, "blabla~pblabla\n", [Args]),
> ok = file:close(IoDescriptor),
> epgsql:squery(C, "SELECT * FROM my_table").
> ```
> 
> etc. Of course, no real files involved, `IoDevice` is just epgsql connection `pid()` that can communicate "io protocol".
> 
> So, you could write arbitrary CSV-formatted data using `file` module's functions.
> 
> Please, let me know if you have other proposals? I don't have much experience with using `COPY`, so, not sure what the most comfortable API would look like.

Seems good to me.

**Comment by dszoboszlay** [comment]
The `IoDescriptor` approach would work fine. If it's easy to implement, go for it!

However, when the data you want to copy in is available as Erlang terms, it could be even more convenient if one could off-load the encoding to `epgsql`. So the API would be something like:

```erlang
ok = epgsql:start_copy(C, "COPY my_table FROM stdin ...", Types = [text, int2]),
ok = epgsql:copy_row(C, Row = [<<"foo">>, 42]),
ok = epgsql:close_copy(C).
```

This would help me with the conversion from Erlang types to CSV - or maybe even the more efficient binary format?

**Comment by seriyps** [comment]
@dszoboszlay the problem is that epgsql is quite good at conversion from Erlang types to postgres binary wire protocol, but it doesn't really know how to convert to/from text representation.
We have optional `decode_text` callback in our codecs

https://github.com/epgsql/epgsql/blob/devel/src/epgsql_codec.erl#L39-L42

but I think none of the codecs implement it. And we don't have `encode_text` at all right now.
However, maybe it makes sense to try to go this way, or have both actually

**Comment by dszoboszlay** [comment]
Yes, I was hoping the binary format of `COPY` is the same thing that is used for parameters of prepared queries, so you could utilise the existing encoders.

**Comment by seriyps** [comment]
While thinking about it, noticed one complication: epgsql currently doesn't have any TCP flow control in it: we use `{active, true}` to receive TCP data from PostgreSQL and we use asynchronous `inet send` instead of `gen_tcp:send`

https://github.com/epgsql/epgsql/blob/95b02f29664bc793d39db3a6576b045dd8b446f7/src/epgsql_sock.erl#L143

https://github.com/epgsql/epgsql/blob/95b02f29664bc793d39db3a6576b045dd8b446f7/src/epgsql_sock.erl#L387-L397

Which might be a big problem for COPY commands, since they tend to produce a lot of data. That means we might need to implement some flow-control first (see, eg, #230)

**Comment by seriyps** [comment]
Made a PR to add support for `COPY ... FROM STDIN` (to transfer data from Erlang to Postgres) https://github.com/epgsql/epgsql/pull/248

`COPY .. TO STDOUT` would require some rework of how we work with TCP socket (need to add some TCP backpressure, because right now we use `{active, true}` and it can easily create an OOM situation when COPY result is very big), see #230.
